File: leetcode/arrays/twosum.cpp
#include<iostream>
#include<vector>
using namespace std;
void input(vector<int>&ivec){
    for(auto &i : ivec){
        cin >> i;
    }
}
void display(vector<int>ivec){
    for(auto const i : ivec){
        cout << i << " ";
    }
    cout << endl;
}
 vector<int> twoSum(vector<int>& nums, int target) {
        for(int i = 0; i < nums.size(); i++){
            //O(N^2) Approach i.e Brute Force
            for(int j = i+1; j < nums.size() ; j++){
                if(nums[j] == target - nums[i]){
                    vector<int>ivec;
                    ivec.push_back(i);
                    ivec.push_back(j);
                    return ivec;
                }
            }
        }
        return nums;
    }
int main()
{   
    int n;
    cin >> n;
    vector<int>nums(n);
    input(nums);
    int target;
    cin >> target;
    display(twoSum(nums, target));
   
    return 0;
}

File: python/hash_map/two_sum.py
def twoSum(nums, target):  
 #Approach 1 using  1 pass Hash Table
#    mydict = {}
#    for i in range(len(nums)):
#        complement = target - nums[i]
#        if complement in mydict:
#            return [i, mydict[complement]]
#        mydict[nums[i]] = i
 #Approach 2 using 2 pass Hash Table
    hashmap = {}
    for i in range(len(nums)):
        hashmap[nums[i]] = i
    for i in range(len(nums)):
        complement = target - nums[i]
        if complement in hashmap and hashmap[complement] !=i:
            return [i, hashmap[complement]]
 #Approach 3 using Brute Force
#    for i in range(len(nums)):
#        for j in range(i+1, len(nums)):
#            if nums[j] == target - nums[i]:
#                return [i,j]

nums = []
[nums.append(int(input())) for _ in range(int(input()))]
target = int(input())
print(twoSum(nums,target))
        
